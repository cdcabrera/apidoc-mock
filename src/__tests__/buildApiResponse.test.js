const buildApiResponse = require('../buildApiResponse');

describe('buildApiResponse', () => {
  it('should return specific properties', () => {
    expect(buildApiResponse).toMatchSnapshot('specific properties');
  });

  it('should get an authorized example', () => {
    expect({
      default: buildApiResponse.getAuthExampleResponse(),
      filteredAndMissing: buildApiResponse.getAuthExampleResponse([
        {
          status: 200
        }
      ]),
      filteredAndFound: buildApiResponse.getAuthExampleResponse([
        {
          status: 200
        },
        {
          status: 401,
          content: 'lorem ipsum dolor sit amet'
        }
      ])
    }).toMatchSnapshot('authorized');
  });

  it('should get an example', () => {
    expect({
      default: buildApiResponse.getExampleResponse(),
      mockForcedAndFilteredAndMissing: buildApiResponse.getExampleResponse({
        mockSettings: { forceStatus: 404 },
        successExamples: undefined,
        errorExamples: []
      }),
      mockForcedExcessiveStatusHigh: buildApiResponse.getExampleResponse({
        mockSettings: { forceStatus: 700 }
      }),
      mockForcedExcessiveStatusLow: buildApiResponse.getExampleResponse({
        mockSettings: { forceStatus: -700 }
      }),
      mockResponseAndFilteredSuccess: buildApiResponse.getExampleResponse({
        mockSettings: { response: 'response' },
        successExamples: [{ status: 201 }],
        errorExamples: []
      }),
      mockResponseAndFilteredError: buildApiResponse.getExampleResponse({
        mockSettings: { response: 'response' },
        successExamples: [],
        errorExamples: [{ status: 400 }]
      }),
      mockSuccessAndFilteredSuccess: buildApiResponse.getExampleResponse({
        mockSettings: { response: 'success' },
        successExamples: [{ status: 201, content: 'lorem ipsum dolor sit amet' }],
        errorExamples: []
      }),
      mockSuccessAndFilteredAndMissing: buildApiResponse.getExampleResponse({
        mockSettings: { response: 'success' },
        successExamples: [],
        errorExamples: [{ status: 400, content: 'lorem ipsum dolor sit amet' }]
      }),
      mockErrorAndFilteredError: buildApiResponse.getExampleResponse({
        mockSettings: { response: 'error' },
        successExamples: [],
        errorExamples: [{ status: 400, content: 'lorem ipsum dolor sit amet' }]
      }),
      mockErrorAndFilteredAndMissing: buildApiResponse.getExampleResponse({
        mockSettings: { response: 'error' },
        successExamples: [{ status: 201, content: 'lorem ipsum dolor sit amet' }],
        errorExamples: []
      })
    }).toMatchSnapshot('example');
  });

  it('should get an authorized and general example', () => {
    const mockFunction = jest.fn();
    buildApiResponse.getMockResponseExample(undefined, {
      getExampleResponse: mockFunction,
      getAuthExampleResponse: mockFunction
    });
    expect(mockFunction).toHaveBeenCalledTimes(2);
  });

  it('should build responses based on examples', () => {
    const mockFunction = jest.fn().mockImplementation(() => jest.fn());

    expect({
      default: buildApiResponse.buildResponses(),
      basic: buildApiResponse.buildResponses({
        getDocs: () => [{ type: 'get', url: 'lorem/ipsum', dolor: 'sit' }],
        getResponseCallback: mockFunction
      })
    }).toMatchSnapshot('responses');

    expect(mockFunction.mock.calls).toMatchSnapshot('getResponseCallback');
    expect(mockFunction).toHaveBeenCalledTimes(1);
  });

  it('should process a request with mock settings and return a response', async () => {
    jest.useFakeTimers();
    const mockRequest = jest.fn();
    const mockRequestWrapper = {
      get: mockRequest
    };

    const mockResponse = jest.fn().mockImplementation((...args) => args);
    const mockResponseWrapper = {
      append: (...args) => mockResponse('append', args),
      end: (...args) => mockResponse('end', args),
      status: (...args) => mockResponse('status', args),
      set: (...args) => mockResponse('set', args),
      send: (...args) => mockResponse('send', args)
    };

    const basicResponse = buildApiResponse.getResponseCallback();
    await basicResponse(mockRequestWrapper, mockResponseWrapper);
    jest.runAllTimers();

    expect(mockResponse.mock.calls).toMatchSnapshot('default response');
    mockResponse.mockClear();
    mockRequest.mockClear();

    const successResponse = buildApiResponse.getResponseCallback({
      successExamples: [{ status: 201, type: 'text/html', content: '<html><body>lorem ipsum dolor sit</body></html>' }]
    });
    await successResponse(mockRequestWrapper, mockResponseWrapper);
    jest.runAllTimers();

    expect(mockResponse.mock.calls).toMatchSnapshot('success response');
    mockResponse.mockClear();
    mockRequest.mockClear();

    const errorResponse = buildApiResponse.getResponseCallback({
      mockSettings: { forceStatus: 400 },
      errorExamples: [{ status: 400, type: 'text/html', content: '<html><body>lorem ipsum dolor sit</body></html>' }]
    });
    await errorResponse(mockRequestWrapper, mockResponseWrapper);
    jest.runAllTimers();

    expect(mockResponse.mock.calls).toMatchSnapshot('error response');
    mockResponse.mockClear();
    mockRequest.mockClear();

    const notAuthorizedResponse = buildApiResponse.getResponseCallback({
      requestHeaders: { Authorization: 'Token AUTH_TOKEN' }
    });
    await notAuthorizedResponse(mockRequestWrapper, mockResponseWrapper);
    jest.runAllTimers();

    expect(mockResponse.mock.calls).toMatchSnapshot('not authorized response');
    mockResponse.mockClear();
    mockRequest.mockClear();

    const authorizedResponse = buildApiResponse.getResponseCallback({
      requestHeaders: { Authorization: 'Token AUTH_TOKEN' }
    });
    await authorizedResponse({ get: (...args) => `mockGetHeader: ${args}` }, mockResponseWrapper);
    jest.runAllTimers();

    expect(mockResponse.mock.calls).toMatchSnapshot('authorized response');
    mockResponse.mockClear();
  });
});
