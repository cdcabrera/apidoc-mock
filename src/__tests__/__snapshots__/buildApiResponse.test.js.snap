// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildApiResponse should build a response with specific properties: buildResponse 1`] = `
{
  "appResponses": [
    {
      "callback": [Function],
      "type": "get",
      "url": "/hello/world/",
    },
  ],
  "routesLoaded": true,
}
`;

exports[`buildApiResponse should build a response with specific properties: buildResponse callback 1`] = `
[
  [
    "Cache-Control",
    "no-cache",
  ],
  [
    200,
  ],
  [
    "Content-Type",
    "text/plain",
  ],
]
`;

exports[`buildApiResponse should build specific headers: buildRequestHeaders.request.headers 1`] = `
{
  "Access-Control-Allow-Credentials": true,
  "Access-Control-Allow-Methods": "GET, PUT, POST, DELETE, HEAD, OPTIONS, PATCH",
  "Access-Control-Allow-Origin": "*",
}
`;

exports[`buildApiResponse should build specific headers: buildRequestHeaders.request.preflight 1`] = `
{
  "Access-Control-Allow-Credentials": true,
  "Access-Control-Allow-Headers": "x-test-header",
  "Access-Control-Allow-Methods": "GET, PUT, POST, DELETE, HEAD, OPTIONS, PATCH",
  "Access-Control-Allow-Origin": "*",
  "access-control-request-headers": "x-test-header",
}
`;

exports[`buildApiResponse should parse custom mock api settings: parseCustomMockSettings 1`] = `
{
  "delay": 2000,
  "forceStatus": 401,
  "reload": true,
  "response": "response",
}
`;

exports[`buildApiResponse should parse malformed mock api settings and provide fallbacks: parseCustomMockSettings.delayResponse.malformed 1`] = `
{
  "delay": 1000,
  "forceStatus": undefined,
  "reload": undefined,
  "response": undefined,
}
`;

exports[`buildApiResponse should parse malformed mock api settings and provide fallbacks: parseCustomMockSettings.forceStatus.malformed 1`] = `
{
  "delay": undefined,
  "forceStatus": 200,
  "reload": undefined,
  "response": undefined,
}
`;

exports[`buildApiResponse should parse random mock api settings: parseCustomMockSettings.randomError 1`] = `
{
  "delay": undefined,
  "forceStatus": undefined,
  "reload": true,
  "response": "error",
}
`;

exports[`buildApiResponse should parse random mock api settings: parseCustomMockSettings.randomSuccess 1`] = `
{
  "delay": undefined,
  "forceStatus": undefined,
  "reload": true,
  "response": "success",
}
`;

exports[`buildApiResponse should return a mock mime type and parsed content: parseContentAndType 1`] = `
{
  "content": ""{\\n  \\"foo\\": \\"hello\\",\\n  \\"bar\\": \\"world\\"\\n}"",
  "contentType": "application/json",
}
`;

exports[`buildApiResponse should return a mock mime type and parsed content: parseContentAndType.fallback 1`] = `
{
  "content": "{
  "foo": "hello",
  "bar": "world"
}",
  "contentType": "text/plain",
}
`;

exports[`buildApiResponse should return a mock mime type and parsed content: parseContentAndType.passthrough 1`] = `
{
  "content": "{
  "foo": "hello",
  "bar": "world"
}",
  "contentType": "lorem/ipsum",
}
`;

exports[`buildApiResponse should return a mock mime type and parsed content: parseContentAndType.svg 1`] = `
{
  "content": "<lorem><ipsum dolor="sit" /></lorem>",
  "contentType": "image/svg+xml",
}
`;

exports[`buildApiResponse should return a mock mime type and parsed content: parseContentAndType.xml 1`] = `
{
  "content": "<lorem><ipsum dolor="sit" /></lorem>",
  "contentType": "text/xml",
}
`;

exports[`buildApiResponse should return specific properties: specific properties 1`] = `
[
  "buildResponse",
  "buildRequestHeaders",
  "getContentAndType",
  "getCustomMockSettings",
  "getExampleResponse",
]
`;
